orc/orcarm.c

```diff
@@ -1016,3 +1016,201 @@ orc_arm64_emit_am (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
+  if (imms == 0x1f)
+    return 0;
+
+  /** must not match UXTx/SXTx alias */
+  if (immr == 0) {
+    if (bits == ORC_ARM64_REG_32) {
+      if (imms == 0x7 || imms == 0xf)
+        return 0;
+    } else if (is_unsigned) {
+      if (imms == 0x7 || imms == 0xf || imms == 0x1f)
+        return 0;
+    }
+  }
+
+  return 1;
+}
+
+void
+orc_arm64_emit_bfm (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
+    int Rn, int Rd, orc_uint32 immr, orc_uint32 imms)
+{
+  orc_uint32 code;
+  int alias = -1;
+
+  static const char *insn_names[3] = {
+    "sbfm", "bfm", "ubfm"
+  };
+  static const char *insn_alias[3][6] = {
+    {"asr","sbfiz","sbfx","sxtb","sxth","sxtw"},
+    {"bfc","bfi","bfxil", "ERROR","ERROR","ERROR"},
+    {"lsl","lsr","ubfiz","ubfx","uxtb","uxth"}
+  };
+
+  char opt_immr[64];
+  char opt_imms[64];
+
+  memset (opt_immr, '\x00', 64);
+  memset (opt_imms, '\x00', 64);
+
+  /** find its alias */
+  switch (opcode) {
+    case ORC_ARM64_DP_SBFM:
+      if (imms == 0x1f) {
+        if (bits == ORC_ARM64_REG_64)
+          imms |= 0x20;
+        sprintf (opt_immr, ", #%u", immr);
+        alias = 0;
+      } else if (imms < immr) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 1;
+      } else if (bfx_preferred(bits, 0, imms, immr)) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 2;
+      } else if (immr == 0) {
+        if (imms == 0x7)
+          alias = 3;
+        else if (imms == 0xf)
+          alias = 4;
+        else if (imms == 0x1f)
+          alias = 5;
+      }
+      break;
+    case ORC_ARM64_DP_BFM:
+      if (imms < immr) {
+        if (Rn == 0x1f) {
+          sprintf (opt_immr, ", #%u", immr);
+          alias = 0;
+        } else {
+          sprintf (opt_immr, ", #%u", immr);
+          sprintf (opt_imms, ", #%u", imms);
+          alias = 1;
+        }
+      } else {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 2;
+      }
+      break;
+    case ORC_ARM64_DP_UBFM:
+      if (imms == 0x1f) {
+        if (bits == ORC_ARM64_REG_64)
+          imms |= 0x20;
+        sprintf (opt_immr, ", #%u", immr);
+        alias = 1;
+      } else if (imms + 1 == immr) {
+        if (bits == ORC_ARM64_REG_64)
+          imms |= 0x20;
+        sprintf (opt_immr, ", #%u", immr);
+        alias = 0;
+      } else if (imms < immr) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 2;
+      } else if (bfx_preferred(bits, 1, imms, immr)) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 3;
+      } else if (immr == 0) {
+        if (imms == 0x7)
+          alias = 4;
+        else if (imms == 0xf)
+          alias = 5;
+      }
+      break;
+    default:
+      ORC_COMPILER_ERROR(p, "unknown opcode %d", opcode);
+      return;
```

Comment:
Our secure-coding checker will complain about the usage of sprintf. Consider using snprintf with the static length (63) for both opt_immr and opt_imms.

Reply:
Got it